src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/Services/DefaultRazorSemanticTokensInfoService.cs

```diff
@@ -80,18 +80,277 @@ internal class DefaultRazorSemanticTokensInfoService : RazorSemanticTokensInfoSe
+                if (requestedRange.End.Line != cachedRange.End.Line)
+                {
+                    var partialRange = new Range { Start = cachedRange.End, End = requestedRange.End };
+                    endTokensTask = ComputeAndCacheTokensAsync(
+                        textDocumentIdentifier, documentSnapshot, documentVersion, semanticVersion,
+                        partialRange, cancellationToken);
+                }
+
+                await Task.WhenAll(startTokensTask, endTokensTask).ConfigureAwait(false);
+
+                var startTokens = await startTokensTask;
+                var endTokens = await endTokensTask;
+
+                return (startTokens, endTokens);
+            }
+
+            async Task<SemanticTokens?> ComputeAndCacheTokensAsync(
+                TextDocumentIdentifier textDocumentIdentifier,
+                DocumentSnapshot documentSnapshot,
+                int documentVersion,
+                VersionStamp semanticVersion,
+                Range partialRange,
+                CancellationToken cancellationToken)
+            {
+                var (tokens, isCSharpFinalized) = await GetSemanticTokensAsync(
+                    textDocumentIdentifier, documentSnapshot, documentVersion, partialRange, cancellationToken).ConfigureAwait(false);
+
+                // Only cache tokens if C# has finalized their tokens (meaning they were computed using
+                // a full semantic model rather than a partial one)
+                if (isCSharpFinalized && tokens is not null)
+                {
+                    _tokensCache.CacheTokens(textDocumentIdentifier.Uri, semanticVersion, partialRange, tokens.Data.ToArray());
+                }
+
+                return tokens;
+            }
+
+            static int AddStartingTokens(
+                Range requestedRange,
+                Range cachedRange,
+                List<int> finalTokens,
+                SemanticTokens startingTokens)
+            {
+                // Add the partially computed range to the results. We need to also keep track of the current
+                // absolute line index to use later when combining results.
+                var absoluteLine = 0;
+                var isFirstTokenInRange = true;
+                for (var tokenIndex = 0; tokenIndex < startingTokens.Data.Length; tokenIndex++)
+                {
+                    // The first int of each token represents the line offset. We use this info to keep track of the
+                    // absolute line for use later on when computing relative token positions.
+                    if (tokenIndex % TokenSize == 0)
+                    {
+                        absoluteLine += startingTokens.Data[tokenIndex];
+
+                        // Skip tokens that are out of the range
+                        if (absoluteLine < requestedRange.Start.Line || absoluteLine >= cachedRange.Start.Line)
+                        {
+                            continue;
```

Comment:
this is super confusing.  Say you have your starting token (at index 0).  So we get into thsi loop.  But then we discover it's line is prior to the range request line.  So we then continue.  then we hit the next int (At index 1) do not get into this if-check, and just add that to the finalTOkens list.

It seems utterly broken.

Reply:
Can you expand on why it seems broken? To explain this check further, sometimes we're passed tokens that fall outside of the requested range. We don't want to include those tokens, so we skip them and don't add them to the final tokens list